Added test of consistency between ckd and mono modes

nollety opened the pull request:
# Description

Add a test that assert that monochromatic results are consistent with CKD results.

We run a one-dimensional experiment with the following settings:

* lambertian surface with reflectance of [0.0, 0.5, 1.0] (parametrised)
* directional illumination with `zenith=50.0 deg` and `azimuth=0.0 deg`
* AFGL (1986) `us_standard` atmosphere (`has_scattering=True`, `has_absorbing=True`)
* `spp=1e4`
* multi-distant measures with 11 viewing zenith angles linearly spaced in [-75, 75]°.

The experiment is run in `mono_double` and `ckd_double` mode with the following specific settings
* `mono_double`:
  * `wavelengths = np.linspace(545.0, 555.0, 1001) * ureg.nm`
* `ckd_double`:
  * `bin_set = "10nm"`
  * `bins = "550"`

Before comparing the results, the monochromatic results are averaged.
The comparaison targets the BRF in the principal plane, for angles inferior to 75° in absolute values.

⚠️ Note that the test will be skipped if the monochromatic absorption cross section data sets for H2O are not present.

The test has also been marked as slow at the moment because it takes around 7 minutes to complete. However, I plan to try to reduce the number of viewing zenith angles and samples per pixel to see if a reasonable executing time is within reach.

Of course, the current spectral coverage (545 to 555 nm) of that test is pretty poor and we shall remedy that.
I'll investigate other spectral regions, but that means that the total executing time of such tests will increase and we will probably end up marking all these tests as slow, which is not necessarily bad I assume.
What do you think? @leroyvn 

# Checklist

- [x] print figures
- [x] more spectral regions
- [x] change surface reflectance
- [x] reduce `n_vza` and `spp`
- [x] number of wavelenght points in mono mode
- [x] add doc-strings
- [x] The code follows the relevant coding guidelines
- [x] The code generates no new warnings
- [x] The code is appropriately documented
- [x] The code is tested to prove its function
- [x] The feature branch is rebased on the current state of the `main` branch
- [x] I give permission that the Eradiate project may redistribute my contributions under the terms of its license

nollety commented:
Below is the BRF comparaison illustrated (note that the monochromatic results are averaged before compared):

![test_onedim_ckd_vs_mono](https://user-images.githubusercontent.com/62285993/140772663-8eddc732-551b-479d-b642-c6f64bae76e6.png)

leroyvn commented on tests/_system/test_onedim_ckd_vs_mono.py:
```diff
@@ -0,0 +1,178 @@
+    ensure_output_dir(os.path.join(output_dir, "plots"))
+    fname_plot = os.path.join(output_dir, "plots", filename)
+
+    mono_results_averaged.brf.squeeze().where(mono_results_averaged.brf.vza < 75).where(
+        mono_results_averaged.brf.vza > -75
```
I think the following is more elegant and should work (please double-check):
```suggestion
    mono_results_averaged.brf.squeeze().where(
        np.logical_and(
            mono_results_averaged.brf.vza < 75, mono_results_averaged.brf.vza > -75
        )
    )
```

nollety replied:
With the vza parametrisation that you suggested, this is not necessary anymore :)

leroyvn commented on tests/_system/test_onedim_ckd_vs_mono.py:
```diff
@@ -0,0 +1,178 @@
+    # Assert averaged mono results are consistent with ckd results
+    mono_brf = (
+        mono_results_averaged.brf.squeeze()
+        .where(mono_results_averaged.brf.vza < 75)
+        .where(mono_results_averaged.brf.vza > -75)
```
ditto

leroyvn commented on tests/_system/test_onedim_ckd_vs_mono.py:
```diff
@@ -0,0 +1,178 @@
+    )
+    ckd_brf = (
+        ckd_results.brf.squeeze()
+        .where(ckd_results.brf.vza < 75)
+        .where(ckd_results.brf.vza > -75)
```
ditto

leroyvn reviewed (commented):
> We run a one-dimensional experiment with the following settings:
>
> * lambertian surface with reflectance of 0.0 (I will change that to a non-zero values, possibly parametrised the test to check with several values)
> * directional illumination with zenith=50.0 deg and azimuth=0.0 deg
> * spp=1e5
> * 40 viewing zenith angles (probably too many)

* You can likely reduce the VZA parametrisation to {-75, -60, -45, -30, -15, 0, 15, 30, 45, 60, 75} (11 values).
* Surface reflectance could be {0, 0.5, 1}.
* What are your atmosphere params?

> The experiment is run in `mono_double` and `ckd` mode with the following specific settings

For consistency, please use the `ckd_double` mode.

> * `mono_double`: `wavelengths=np.linspace(545.0, 555.0, 2) * ureg.nm`
> * `ckd`: `bin_set="10nm", bins="550"`

I'm surprised that this yields correct results at all: I guess this means that absorption plays a diminutive role, since two wavelengths in mono mode are insufficient so as to capture absorption features correctly. IMO you don't even have to bother running two monochromatic simulations and averaging them: just simulate once at 550 nm.

> Note that the test will be skipped if the monochromatic absorption cross section data sets for H2O are not present.

Could we ship a minimal dataset for testing purposes or is this too much data?

> The test has also been marked as slow at the moment because it takes 40 seconds to complete. However, I plan to try to reduce the number of viewing zenith angles and samples per pixel to see if a reasonable executing time is within reach.

Good.

> Of course, the current spectral coverage (545 to 555 nm) of that test is pretty poor and we shall remedy that. I'll investigate other spectral regions, but that means that the total executing time of such tests will increase and we will probably end up marking all these tests as slow, which is not necessarily bad I assume. What do you think?

Slow tests are fine, but they still have to mean something. In its current form, is this test really checking that the CKD mode yields correct results? Sort of, but only if you consider that it's about testing that the basic machinery works as expected.

A proper assessment of the accuracy of our CKD method is a more ambitious test because it requires running a complete spectral simulation on a band—way out of the scope of the test suite.

nollety commented:
>     * You can likely reduce the VZA parametrisation to {-75, -60, -45, -30, -15, 0, 15, 30, 45, 60, 75} (11 values).
> 

Done!

>     * Surface reflectance could be {0, 0.5, 1}.

Done!

>     * What are your atmosphere params?

I've updated the PR description with the atmosphere parameters.
 
> > The experiment is run in `mono_double` and `ckd` mode with the following specific settings
> 
> For consistency, please use the `ckd_double` mode.

Right, this is done now.

> > * `mono_double`: `wavelengths=np.linspace(545.0, 555.0, 2) * ureg.nm`
> > * `ckd`: `bin_set="10nm", bins="550"`
> 
> I'm surprised that this yields correct results at all: I guess this means that absorption plays a diminutive role, since two wavelengths in mono mode are insufficient so as to capture absorption features correctly. IMO you don't even have to bother running two monochromatic simulations and averaging them: just simulate once at 550 nm.

This is a copy-paste mistake, I've run the test with different number of wavelengths in monochromatic mode and ended up using only 2 when experimenting with the scene parameters.

> Could we ship a minimal dataset for testing purposes or is this too much data?

I think there is room for the required datasets on our FTP server, but not on `eradiate-data`.
 
> A proper assessment of the accuracy of our CKD method is a more ambitious test because it requires running a complete spectral simulation on a band—way out of the scope of the test suite.

I changed the number of wavelengths in monochromatic mode to 1001 which correspond to a 0.01 nm resolution, i.e. 10 times less than the native datasets resolution. I think that if we work in [549.5, 550.5] nm with the bin "550" from the "1nm" bin set (once the datasets are ready, ETA: beginning next week), we could achieve that.

nollety commented:
I've duplicated the test for the [1045, 1055] nm wavelength bin, just to confirm that things work well there too (and it does). I don't believe it's worth it to add any more wavelength bin for the time being.

nollety commented:
We can always add more tests of the kind later.

nollety commented:
Below are the figures that the tests output (RAD stands for "relative absolute difference"):

![test_onedim_ckd_vs_mono_550_0 0](https://user-images.githubusercontent.com/62285993/141153575-54ce8125-b3bf-4282-b5c2-9ea94776c335.png)

![test_onedim_ckd_vs_mono_550_0 5](https://user-images.githubusercontent.com/62285993/141154174-08f8e909-4a9b-4cc0-ad0f-fe48d7549fef.png)

![test_onedim_ckd_vs_mono_550_1 0](https://user-images.githubusercontent.com/62285993/141155402-113daa23-8213-416f-a968-8d3253baa6ea.png)

![test_onedim_ckd_vs_mono_1050_0 0](https://user-images.githubusercontent.com/62285993/141155421-6ef51efa-8553-4034-8a02-f1484b911519.png)

![test_onedim_ckd_vs_mono_1050_0 5](https://user-images.githubusercontent.com/62285993/141155998-578c0d91-613b-4d98-8e3a-f9efb5087f35.png)

![test_onedim_ckd_vs_mono_1050_1 0](https://user-images.githubusercontent.com/62285993/141156567-ae449003-594f-4122-aadc-84a5b22b8836.png)

leroyvn commented on tests/_system/test_onedim_ckd_vs_mono.py:
```diff
@@ -0,0 +1,310 @@
+    brf_mono_averaged.plot(x="vza", ls="dotted", marker=".")
+    brf_ckd.plot(x="vza", ls="dotted", marker=".")
+
+    # Compute relative absolute difference and annotate plot with min, max and
+    # mean values
```
I find the "relative absolute difference" nomenclature to be quite confusing in our context ("absolute bias" is actually a signed quantity in usual EO terminology). However, since this is correct vocabulary, let's clarify this:
```suggestion
    # Compute relative absolute difference and annotate plot with min, max and
    # mean values
    # NB: RAD(x, y) = abs(x - y) / x
```
Please also add similar detail in the text docstring.

nollety replied:
I understand your confusion. I derived the definition of "relative absolute difference" from the "mean absolute difference" definition from [wikipedia](https://en.wikipedia.org/wiki/Mean_absolute_difference#Relative_mean_absolute_difference) and it sounds a little weird as is 😃.
Where did you find the definition of absolute bias?
It looks like it would be nice to settle this vocabulary throughout eradiate since we use it a lot, don't you think?
As defined, the "relative absolute difference" can be negative, when `x` is negative (unlikely given that the quantities that we usually compare are defined positive).

leroyvn reviewed (changes requested):
Hi @nollety, seems good to me apart from a few details. When these are taken care of, we can merge!

nollety commented on tests/_system/test_onedim_ckd_vs_mono.py:
```diff
@@ -0,0 +1,310 @@
+    # Assert averaged mono results are consistent with ckd results
+    mono_brf = mono_results_averaged.brf.squeeze().values
+    ckd_brf = ckd_results.brf.squeeze().values
+    assert np.allclose(mono_brf, ckd_brf, rtol=5e-2, equal_nan=True)
```
This is probably not required anymore. I was expecting nan values as a result of applying `where(vza < 75)` to the BRF data arrays ; the values corresponding to vza >= 75° would be set to nan, but these values were out of the scope of the comparison, hence the `equal_nan=True`. I'll remove this setting.

leroyvn commented on tests/_system/test_onedim_ckd_vs_mono.py:
```diff
@@ -0,0 +1,375 @@
+      wavelength bin
+
+    Results
+    -------
```
I suggest also annotating the plots
```suggestion
    -------
    
    Note: :math:`\mathrm{RAD}(x, y) = \left| \dfrac{x - y}{x} \right|`

```

leroyvn commented on tests/_system/test_onedim_ckd_vs_mono.py:
```diff
@@ -0,0 +1,375 @@
+
+    Results
+    -------
+
```
Ditto:
```suggestion
    
    Note: :math:`\mathrm{RAD}(x, y) = \left| \dfrac{x - y}{x} \right|`

```

leroyvn reviewed (commented):
Hi @nollety, let's also annotate the plots (the LaTeX might require a very quick check). Then let's merge!

nollety commented:
For the record, these tests are very slow:

* tests run time **without** these tests: 2 minutes 10 seconds
* tests run time **including** these tests: ~23 minutes

nollety commented:
@leroyvn I've added the annotation you suggested and checked the LaTeX rendering, this is ready to be merged! 😃

leroyvn commented:
Thanks! Let's go :)
